brush-core/src/builtins/mapfile.rs

```diff
@@ -93,46 +80,71 @@ impl builtins::Command for MapFileCommand {
     ) -> Result<variables::ArrayLiteral, error::Error> {
-        let mut entries = vec![];
-
         let orig_term_attr = setup_terminal_settings(&input_file)?;
 
-        let mut current_entry = String::new();
-        let mut buffer: [u8; 1] = [0; 1]; // 1-byte buffer
+        let mut entries = vec![];
+        let mut line = vec![];
+        let mut idx = self.skip_count;
+        let mut read_count = 0;
+        let delimiter = self.delimiter.chars().next().unwrap_or('\n') as u8;
+
+        let mut reader = input_file.clone();
+        let mut buf = [0u8; 1];
+
+        while self.max_count == 0 || entries.len() < usize::try_from(self.max_count)? {
+            line.clear();
+            let mut saw_delimiter = false;
+
+            loop {
+                match reader.read(&mut buf) {
+                    Ok(0) => break,                                         // End of input
+                    Ok(1) if buf[0] == b'\x03' => break,                    // Ctrl+C
+                    Ok(1) if buf[0] == b'\x04' && line.is_empty() => break, // Ctrl+D
+                    Ok(1) => {
+                        let byte = buf[0];
+                        line.push(byte);
+                        if byte == delimiter {
+                            saw_delimiter = true;
+                            break;
+                        }
+                    }
+                    Ok(_) => unreachable!("input can only be 0, 1, or error"),
+                    Err(e) => return Err(e.into()),
+                }
+            }
 
-        loop {
-            // TODO: Figure out how to restore terminal settings on error?
-            let n = input_file.read(&mut buffer)?;
-            if n == 0 {
-                // EOF reached.
+            if line.is_empty() && !saw_delimiter {
                 break;
             }
 
-            let ch = buffer[0] as char;
+            if read_count < self.skip_count {
+                read_count += 1;
+                continue;
+            }
 
-            // Check for Ctrl+C.
-            if ch == '\x03' {
-                break;
-            // Ctrl+D is EOF *if* there's no entry in progress.
-            } else if ch == '\x04' && current_entry.is_empty() {
-                break;
+            if self.remove_delimiter && line.ends_with(&[delimiter]) {
+                line.pop();
             }
 
-            // Check for a delimiting newline char.
-            // TODO: Support other delimiters.
-            if ch == '\n' {
-                if !self.remove_delimiter {
-                    current_entry.push(ch);
-                }
+            let line_str = String::from_utf8_lossy(&line).to_string();
 
-                entries.push((None, std::mem::take(&mut current_entry)));
-            } else {
-                current_entry.push(ch);
+            if let Some(callback) = &self.callback {
+                if (idx - self.origin.unwrap_or(0)) % self.callback_group_size == 0 {
```

Comment:
Is there a risk of division by zero if `self.callback_group_size` is `0`? Not sure how `bash` handles that.

Reply:
It throws an invalid callback quantum error: https://github.com/bminor/bash/blob/6794b5478f660256a1023712b5fc169196ed0a22/builtins/mapfile.def#L259